### PR TITLE
[`Llama2`]  Add support for Llama 2

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Add Support for Llama2 ! 🔥 🔥

### CODE REVIEW

File: tests/pipelines/test_pipelines_conversational.py [hunk 1]
```diff
@@ -216,6 +216,40 @@ def test_integration_torch_conversation_dialogpt_input_ids(self):
...
+        self.assertEqual(inputs, EXPECTED_INPUTS_IDS)
+
+        model = AutoModelForCausalLM.from_pretrained("llamaste/Llama-2-7b-chat-hf")
+        conversation_agent = ConversationalPipeline(model=model, tokenizer=tokenizer)
+
+        conversation = Conversation(
+            "<<SYS>>\n Only answer with emojis, and charades\n<</SYS>>\n\nHow can I build a house in 10 steps?"
+        )
+        result = conversation_agent(conversation)
+        self.assertEqual(result.generated_responses[-1], EXPECTED_INPUTS_IDS)
```
Comment (ArthurZucker) [inline tests/pipelines/test_pipelines_conversational.py:251]: Just need to update with expected values here

File: tests/models/llama/test_modeling_llama.py [hunk 2]
```diff
@@ -365,3 +365,193 @@ def test_model_rope_scaling(self, scaling_type):
...
+    @slow
+    def test_model_7b_logits(self):
+        input_ids = [1, 306, 4658, 278, 6593, 310, 2834, 338]
+        model = LlamaForCausalLM.from_pretrained("/raid/arthur/llama-7b", device_map="auto")
+        out = model(torch.tensor(input_ids))
+        # Expected mean on dim = -1
+        EXPECTED_MEAN = torch.tensor([[-6.6550, -4.1227, -4.9859, -3.2406, 0.8262, -3.0033, 1.2964, -3.3699]])
+        torch.testing.assert_close(out.mean(-1), EXPECTED_MEAN, atol=1e-2, rtol=1e-2)
+        # slicing logits[0, 0, 0:30]
+        EXPECTED_SLICE = torch.tensor(
```
Comment (ArthurZucker) [inline tests/models/llama/test_modeling_llama.py:381]: Format slices correctluy

File: docs/source/en/model_doc/llama.md [hunk 3]
```diff
@@ -50,6 +50,12 @@ come in several checkpoints they each contain a part of each weight of the model
 
 This model was contributed by [zphang](https://huggingface.co/zphang) with contributions from [BlackSamorez](https://huggingface.co/BlackSamorez). The code of the implementation in Hugging Face is based on GPT-NeoX [here](https://github.com/EleutherAI/gpt-neox). The original code of the authors can be found [here](https://github.com/facebookresearch/llama).
 
+
+Based on the original Llama model, Meta AI has released some follow-up works:
```
Comment (sgugger) [inline docs/source/en/model_doc/llama.md:54]:
```suggestion
Based on the original LLaMA model, Meta AI has released some follow-up works:
```

File: docs/source/en/model_doc/llama2.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+The abstract from the paper is the following:
+
+*In this work, we develop and release Llama 2, a collection of pretrained and fine-tuned large language models (LLMs) ranging in scale from 7 billion to 70 billion parameters. Our fine-tuned LLMs, called Llama 2-Chat, are optimized for dialogue use cases. Our models outperform open-source chat models on most benchmarks we tested, and based on our human evaluations for helpfulness and safety, may be a suitable substitute for closed- source models. We provide a detailed description of our approach to fine-tuning and safety improvements of Llama 2-Chat in order to enable the community to build on our work and contribute to the responsible development of LLMs.*
+
+Tips:
+
+- Weights for the Llama2 models can be obtained from by filling out [this form](TO COME)
+- The architecture is very similar to the first Llama, with the addition of Groupe Query Attention (GQA) following this [paper](https://arxiv.org/pdf/2305.13245.pdf)
+- Setting `config.pretraining_tp` to a value different than 1 will acctivatethe more accurate but slower computation of the linear layers, which should better match the original logits.
```
Comment (sgugger) [inline docs/source/en/model_doc/llama2.md:31]:
```suggestion
- Setting `config.pretraining_tp` to a value different than 1 will activate the more accurate but slower computation of the linear layers, which should better match the original logits.
```

File: docs/source/en/model_doc/llama2.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+- The original model uses `pad_id = -1` which means that there is not padding token. We can't have the same logic, so we decide to pad with `0` by default. Then, the `model.config.pad_token_id` was properly set with the padding index. The `embed_tokens` layer of the model is initialized with`self.embed_tokens = nn.Embedding(config.vocab_size, config.hidden_size, self.config.padding_idx)`, which makes sure that encoding the padding token will output zeros.
+
+- After downloading the weights, they will need to be converted to the Hugging Face Transformers format using the [conversion script](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/convert_llama_weights_to_hf.py). The script can be called with the following (example) command:
```
Comment (sgugger) [inline docs/source/en/model_doc/llama2.md:34]: Is there no system were people having access can download directly from the Hub as long as they put the same email address as their HF account in the form?

File: src/transformers/models/llama/tokenization_llama.py [hunk 6]
```diff
@@ -46,10 +49,24 @@
...
+If a question does not make any sense, or is not factually coherent, explain why instead of answering something not\
+correct. If you don't know the answer to a question, please don't share false information."""
+# fmt: on
+
 
 class LlamaTokenizer(PreTrainedTokenizer):
     """
-    Construct a Llama tokenizer. Based on byte-level Byte-Pair-Encoding.
+    Construct a Llama tokenizer. Based on byte-level Byte-Pair-Encoding. The padding token is unset as there is not
+    padding token in the original model. Inputs are padded using `-1`, which is similar to masking the inputs with
```
Comment (sgugger) [inline src/transformers/models/llama/tokenization_llama.py:68]: Passing value needs updating.

**Review by sgugger — COMMENTED** [review]
Thanks a lot for adding this! I have two commens: let's double-check the casing on the model name and the padding 0 value (I'm afraid the old 0 token is getting its embedding erased)

File: docs/source/en/_toctree.yml [hunk 7]
```diff
@@ -335,6 +335,8 @@
         title: LED
       - local: model_doc/llama
         title: LLaMA
+      - local: model_doc/llama2
+        title: Llama2
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply docs/source/en/_toctree.yml:339]: It was requested by the Llama Team!

File: docs/source/en/model_doc/llama2.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+Tips:
+
+- Weights for the Llama2 models can be obtained from by filling out [this form](TO COME)
+- The architecture is very similar to the first Llama, with the addition of Groupe Query Attention (GQA) following this [paper](https://arxiv.org/pdf/2305.13245.pdf)
+- Setting `config.pretraining_tp` to a value different than 1 will acctivatethe more accurate but slower computation of the linear layers, which should better match the original logits.
+- The original model uses `pad_id = -1` which means that there is not padding token. We can't have the same logic, so we decide to pad with `0` by default. Then, the `model.config.pad_token_id` was properly set with the padding index. The `embed_tokens` layer of the model is initialized with`self.embed_tokens = nn.Embedding(config.vocab_size, config.hidden_size, self.config.padding_idx)`, which makes sure that encoding the padding token will output zeros.
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply docs/source/en/model_doc/llama2.md:32]: Yes, indeed, we should probably just go with None...

File: docs/source/en/model_doc/llama2.md (hunk 5 above)
Comment (ArthurZucker) [reply docs/source/en/model_doc/llama2.md:34]: Actually yes 🤣 they should be able to access all of them nice catch

File: src/transformers/models/llama/convert_llama_weights_to_hf.py [hunk 9]
```diff
@@ -59,17 +58,22 @@
...
     "30B": 4,
     "65B": 8,
+    "70B": 8,
+    "70Bf": 8,
 }
 
 
-def compute_intermediate_size(n):
-    return int(math.ceil(n * 8 / 3) + 255) // 256 * 256
+def compute_intermediate_size(n, ffn_dim_multiplier=1, multiple_of=256):
```
Comment (AgentJ-WR) [inline src/transformers/models/llama/convert_llama_weights_to_hf.py:75]:
```suggestion
def compute_intermediate_size(n, ffn_dim_multiplier=1, multiple_of=256):
    """
    Compute the size of an intermediate layer for a feed-forward network (ffn).

    Parameters:
    n (int): The model size.
    ffn_dim_multiplier (float): Multiplier for the intermediate layer size relative to `n`.
    multiple_of (int): The size is rounded up to the nearest multiple of this value.

    Returns:
    int: The computed size of the intermediate layer.
    """

    # Input validation
    if not isinstance(n, int) or not isinstance(multiple_of, int) or not isinstance(ffn_dim_multiplier, (int, float)):
        raise TypeError("Invalid input type. Expected types: n (int), ffn_dim_multiplier (int/float), multiple_of (int).")
    if n < 0 or ffn_dim_multiplier < 0 or multiple_of <= 0:
        raise ValueError("Invalid input values. Both `n` and `ffn_dim_multiplier` should be non-negative, and `multiple_of` should be greater than 0.")

    # Calculation
    return multiple_of * ((int(ffn_dim_multiplier * int(8 * n / 3)) + multiple_of - 1) // multiple_of)

```

**Review by sgugger — APPROVED** [review]

File: src/transformers/models/llama/convert_llama_weights_to_hf.py [hunk 10]
```diff
@@ -259,18 +281,20 @@ def main():
     )
     parser.add_argument(
         "--model_size",
-        choices=["7B", "13B", "30B", "65B", "tokenizer_only"],
+        choices=["7B", "7Bf", "13B", "13Bf", "30B", "65B", "70B", "70Bf", "tokenizer_only"],
```
Comment (shijie-wu) [inline src/transformers/models/llama/convert_llama_weights_to_hf.py:284]: what does the `f` suffix mean?

File: src/transformers/models/llama/modeling_llama.py [hunk 11]
```diff
@@ -304,10 +352,15 @@ def forward(
...
 
-        attn_output = attn_output.transpose(1, 2)
+        attn_output = attn_output.transpose(1, 2).contiguous()
         attn_output = attn_output.reshape(bsz, q_len, self.hidden_size)
 
-        attn_output = self.o_proj(attn_output)
+        if self.pretraining_tp > 1:
+            attn_output = attn_output.split(self.hidden_size // self.pretraining_tp, dim=2)
+            o_proj_slices = self.o_proj.weight.split(self.hidden_size // self.pretraining_tp, dim=1)
+            attn_output = sum([F.linear(attn_output[i], o_proj_slices[i]) for i in range(self.pretraining_tp)])
```
Comment (Atry) [inline src/transformers/models/llama/modeling_llama.py:361]: @ArthurZucker Shouldn't it be `torch.cat`?

File: README.md [hunk 12]
```diff
@@ -386,6 +386,7 @@ Current number of checkpoints: ![](https://img.shields.io/endpoint?url=https://h
...
...ased with the paper [Llama2: Open Foundation and Fine-Tuned Chat Models](https://ai.meta.com/research/publications/llama-2-open-foundation-and-fine-tuned-chat-models/XXX) by Hugo Touvron, Louis Martin, Kevin Stone, Peter Albert, Amjad Almahairi, Yasmine Babaei, Nikolay Bashlykov, Soumya Batra, Prajjwal Bhargava, Shruti Bhosale, Dan Bikel, Lukas Blecher, Cristian Canton Ferrer, Moya Chen, Guillem Cucurull, David Esiobu, Jude Fernandes, Jeremy Fu, Wenyin Fu, Brian Fuller, Cynthia Gao, Vedanuj Goswami, Naman Goyal, Anthony Hartshorn, Saghar Hosseini, Rui Hou, Hakan Inan, Marcin Kardas, Viktor Kerkez Madian Khabsa, Isabel Kloumann, Artem Korenev, Punit Singh Koura, Marie-Anne Lachaux, Thibaut Lavril, Jenya Lee, Diana Liskovich, Yinghai Lu, Yuning Mao, Xavier Martinet, Todor Mihaylov, Pushka rMishra, Igor Molybog, Yixin Nie, Andrew Poulton, Jeremy Reizenstein, Rashi Rungta, Kalyan Saladi, Alan Schelten, Ruan Silva, Eric Michael Smith, Ranjan Subramanian, Xiaoqing EllenTan, Binh Tang, Ross Taylor, Adina Williams, Jian Xiang Kuan, Puxin Xu, Zheng Yan, Iliyan Zarov, Yuchen Zhang, Angela Fan, Melanie Kambadur, Sharan Narang, Aurelien Rodriguez, Robert Stojnic, Sergey Edunov, Thomas Scialom.
```
Comment (ayaka14732) [inline README.md:389]: "Llama2" should be "Llama 2"

**Comment by fe1ixxu** [comment]
Somehow this commit will easily lead the llama model to be overflow in fp16 during training. Hope someone can take a look :)

**Comment by ArthurZucker** [comment]
@fe1ixxu you can have a look at #25065 for more details on this!
